### PR TITLE
Add ClusterMetadataService to track and replicate changes to DATA node membership

### PR DESCRIPTION
**Pull request by kuujo**
This PR adds a new `ClusterMetadataService` used to track and replicate the `DATA` node membership in the cluster. The service uses an anti-entropy protocol to replicate the membership and provides event listeners to react to changes in the set of data nodes.

### CODE REVIEW

File: cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import io.atomix.cluster.NodeId;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Cluster metadata anti-entropy advertisement.
+ */
+final class ClusterMetadataAdvertisement {
```
Comment (johnou) [inline cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java]: shouldn't this be serializable?

File: cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    this.digests = digests;
+  }
+
+  /**
+   * Returns the set of digests in the advertisement.
+   *
+   * @return the set of digests in the advertisement
+   */
+  public Set<NodeId> digests() {
+    return digests.keySet();
```
Comment (johnou) [inline cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java]: be careful of this pitfall with kryo https://github.com/magro/kryo-serializers/issues/81

Comment (kuujo) [reply cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java]: That's interesting. And it feels familiar

File: cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java (hunk 1 above)
Comment (kuujo) [reply cluster/src/main/java/io/atomix/cluster/impl/ClusterMetadataAdvertisement.java]: All the appropriate serializers needed for Kryo to handle the serialization are registered.

**Pull request merged**